tests/net/lib/mqtt_sn_client/src/mqtt_sn_client.c

```diff
@@ -122,24 +148,175 @@ static void mqtt_sn_connect_no_will(struct mqtt_sn_client *client)
 	zassert_equal(err, 0, "unexpected error %d");
 	zassert_true(tp_initialized, "Transport not initialized");
 
+	err = mqtt_sn_add_gw(client, gw_id, gw_addr);
+	zassert_equal(err, 0, "unexpected error %d");
+	zassert_equal(evt_cb_data.called, 0, "Unexpected event");
+	zassert_false(sys_slist_is_empty(&client->gateway), "GW not saved.");
+
 	err = mqtt_sn_connect(client, false, false);
 	zassert_equal(err, 0, "unexpected error %d");
-	assert_msg_send(1, 12);
+	assert_msg_send(1, 12, &gw_addr);
 	zassert_equal(client->state, 0, "Wrong state");
 	zassert_equal(evt_cb_data.called, 0, "Unexpected event");
 
-	err = input(client, connack, sizeof(connack));
+	err = input(client, connack, sizeof(connack), &gw_addr);
 	zassert_equal(err, 0, "unexpected error %d");
 	zassert_equal(client->state, 1, "Wrong state");
 	zassert_equal(evt_cb_data.called, 1, "NO event");
 	zassert_equal(evt_cb_data.last_evt.type, MQTT_SN_EVT_CONNECTED, "Wrong event");
 	k_sleep(K_MSEC(10));
 }
 
-static ZTEST(mqtt_sn_client, test_mqtt_sn_connect_no_will)
+static ZTEST(mqtt_sn_client, test_mqtt_sn_handle_advertise)
+{
+	static uint8_t advertise[] = {5, 0x00, 0x0c, 0, 3};
+	static uint8_t connack[] = {3, 0x05, 0x00};
+	int err;
+
+	err = mqtt_sn_client_init(mqtt_client, &client_id, &transport, evt_cb, tx, sizeof(tx), rx,
+				  sizeof(rx));
+	zassert_equal(err, 0, "unexpected error %d");
+
+	err = input(mqtt_client, advertise, sizeof(advertise), &gw_addr);
+	zassert_equal(err, 0, "unexpected error %d");
+	zassert_false(sys_slist_is_empty(&mqtt_client->gateway), "GW not saved.");
+	zassert_equal(evt_cb_data.called, 1, "NO event");
+	zassert_equal(evt_cb_data.last_evt.type, MQTT_SN_EVT_ADVERTISE, "Wrong event");
+	k_sleep(K_SECONDS(1));
+
+	err = input(mqtt_client, advertise, sizeof(advertise), &gw_addr);
+	zassert_equal(err, 0, "unexpected error %d");
+	zassert_false(sys_slist_is_empty(&mqtt_client->gateway), "GW not saved.");
+	zassert_equal(sys_slist_len(&mqtt_client->gateway), 1, "Too many Gateways stored.");
+	zassert_equal(evt_cb_data.called, 2, "Unexpected event");
+	zassert_equal(evt_cb_data.last_evt.type, MQTT_SN_EVT_ADVERTISE, "Wrong event");
+
+	err = mqtt_sn_connect(mqtt_client, false, false);
+	zassert_equal(err, 0, "unexpected error %d");
+	assert_msg_send(1, 12, &gw_addr);
+	zassert_equal(mqtt_client->state, 0, "Wrong state");
+	zassert_equal(evt_cb_data.called, 2, "Unexpected event");
+
+	err = input(mqtt_client, connack, sizeof(connack), &gw_addr);
+	zassert_equal(err, 0, "unexpected error %d");
+	zassert_equal(mqtt_client->state, 1, "Wrong state");
+	zassert_equal(evt_cb_data.called, 3, "NO event");
+	zassert_equal(evt_cb_data.last_evt.type, MQTT_SN_EVT_CONNECTED, "Wrong event");
+
+	k_sleep(K_SECONDS(10));
```

Comment:
The various sleeps in the tests are problematic. These just increase the CI pass time. Would it be possible to have for example a semaphore that waits, but is then released when the condition passes?

Reply:
Done! I removed all sleeps and replaced with semaphores. I also reduced the configured wait time to reduce test idle time.